app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,178 @@
+     * complete).
+     * If the number of tiles exceeds maxTiles, only the first maxSize tiles are cached.
+     */
+    fun replaceAllInBBox(items: Collection<T>, bbox: BoundingBox) = synchronized(this) {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val (completelyContainedTiles, incompleteTiles) = tiles.partition { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
+        if (incompleteTiles.isNotEmpty()) {
+            Log.w(TAG, "bbox does not align with tiles, clearing incomplete tiles from cache")
+            for (tile in incompleteTiles) {
+                removeTile(tile)
+            }
+        }
+        replaceAllInTiles(items, completelyContainedTiles)
+
+        trim()
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(items: Collection<T>, tiles: Collection<TilePos>) {
+        // create / replace tiles
+        for (tile in tiles) {
+            removeTile(tile)
+            byTile[tile] = HashSet()
+        }
+/*        for (item in items) {
+            val tile = byTile[item.getTilePos()] ?: continue
+            tile.add(item)
+            byKey[item.getKey()] = item
+        }*/
+        val bboxByTile = byTile.keys.associateWith { it.asBoundingBox(tileZoom) }
+        items@for (item in items) {
+            val pos = item.getPosition()
+            for (tile in tiles) {
+                if (pos in bboxByTile[tile]!!) {
+                    byTile[tile]!!.add(item)
+                    byKey[item.getKey()] = item
+                    continue@items
+                }
+            }
+        }
+    }
+
+    /** @return all items inside [bbox], items will be loaded if necessary via [fetch] */
+    fun get(bbox: BoundingBox): List<T> = synchronized(this) {
+        val requiredTiles = bbox.asListOfEnclosingTilePos()
+
+        val tilesToFetch = requiredTiles.filterNot { byTile.containsKey(it) }
+        val tilesRect = tilesToFetch.minTileRect()
+        if (tilesRect != null) {
+            val newItems = fetch(tilesRect.asBoundingBox(tileZoom))
+            replaceAllInTiles(newItems, tilesToFetch)
+        }
+
+        val items = requiredTiles.flatMap { tile ->
+            if (tile.asBoundingBox(tileZoom).isCompletelyInside(bbox))
+                byTile[tile]!!
+            else
+                byTile[tile]!!.filter { it.getPosition() in bbox }
+        }
+
+        trim()
+
+        return items
+    }
+
+    /** Reduces cache size to the given number of non-empty [tiles]. */
+    fun trim(tiles: Int = maxTiles) = synchronized(this) {
+        // Empty tiles are kept, as the barely use memory. This avoids empty tiles pushing other
+        // tiles out of the cache and thus may avoid database fetches.
+        if (byTile.count { it.value.isNotEmpty() } <= tiles) return
+
+        while (byTile.count { it.value.isNotEmpty() } > tiles) {
+            removeTile(byTile.entries.firstOrNull { it.value.isNotEmpty() }?.key ?: byTile.keys.first())
+        }
```

Comment:
I find this a bit hard to read

Reply:
changed to
```kotlin
            val firstNonEmptyTile = byTile.entries.firstOrNull { it.value.isNotEmpty() }?.key ?: return
            removeTile(firstNonEmptyTile)
```
(also slightly changes behavior, so that empty tiles are never removed as part of `trim`)